data_pipeline/generic_web_api/helper.py

```diff
@@ -0,0 +1,181 @@
+        self.date_format = date_format
+        self.next_page_cursor = next_page_cursor
+        self.page_number_param = page_number_param
+        self.offset_param = offset_param
+        self.page_size_param = page_size_param
+        self.page_size = page_size
+        self.compose_able_url_key_val = compose_able_url_key_val
+        self.type_specific_params = kwargs
+
+    def _get_url_separator(self):
+        url = self.url_excluding_configurable_parameters
+        if "?" in url:
+            if url.strip().endswith("&") or url.strip().endswith("?"):
+                url_separator = ""
+            else:
+                url_separator = "&"
+        else:
+            url_separator = "?"
+        return url_separator
+
+    def compose_url(self, parameters_key_value: dict):
+        url = self.url_excluding_configurable_parameters
+        url_separator = self._get_url_separator()
+        params = parse.urlencode(
+            {
+                key: value
+                for key, value in parameters_key_value.items() if key and value
+            }
+        )
+
+        return url + url_separator + params
+
+    def get_url(
+            self,
+            url_compose_param: UrlComposeParam,
+    ):
+        start_date = datetime_to_string(
+            url_compose_param.from_date, self.date_format
+        )
+        end_date = datetime_to_string(
+            url_compose_param.to_date, self.date_format
+        )
+        param_dict = dict((key, value) for key, value in [
+            (self.from_date_param, start_date),
+            (self.next_page_cursor, url_compose_param.cursor),
+            (self.to_date_param, end_date),
+            (self.page_number_param, url_compose_param.page_number),
+            (self.offset_param, url_compose_param.page_offset),
+            (
+                self.page_size_param,
+                url_compose_param.page_size or self.page_size
+            )
+            ] if key and value)
+        param_dict = {
+            **param_dict,
+            **self.compose_able_url_key_val
+        }
+
+        return self.compose_url(param_dict)
+
+
+def get_url_manager(url_source_type: str = None):
+    manager = DynamicURLManager
+
+    if url_source_type.strip().lower() == 'civi':
+        manager = DynamicCiviURLManager
+    return manager
+
+
+class DynamicCiviURLManager(DynamicURLManager):
+
+    def get_url(
+            self,
+            url_compose_param: UrlComposeParam,
+    ):
+        start_date = datetime_to_string(
+            url_compose_param.from_date, self.date_format
+        )
+        options = dict((key, value) for key, value in [
+            (self.offset_param, url_compose_param.page_offset),
+            (
+                self.page_size_param,
+                url_compose_param.page_size or self.page_size
+            )
+            ] if key and value)
+        start_date_param = {
+            self.from_date_param: {">=": start_date}
```

Comment:
this seems to end up as a nested dict. e.g.

```json
{
  'modified_date': {
    '>=': '2020-01-01'
  }
}
```

I believe that will be equivalent to (using the string representation of the value / dict):

```json
{
  'modified_date': "{'>=': '2020-01-01'}"
}
```

If that is correct, this should probably explicitly formatted.

Reply:
Apologies, I see now that this is formatted using `json.dumps` and passed as the `json` parameter. Maybe that could be reflected in the variable name (`url_query_param` would suggest to me regular url query parameters)

Reply:
nested dict is right
check out the api call below
```
https://crm.elifesciences.org/crm/sites/all/modules/civicrm/extern/rest.php?entity=Contact&action=get&api_key=userkey&key=sitekey&json={"sequential":1,"modified_date":{">=":"some_date"},"options":{"limit":100,"offset":2000}}
```

Reply:
Is there any documentation that describes that?

Reply:
> Is there any documentation that describes that?

as far as I know, there is no documentation that describe that, one needs to figure it out by exploring the api using api explorer page of your civi crm deployment

Reply:
> > Is there any documentation that describes that?
> 
> as far as I know, there is no documentation that describe that, one needs to figure it out by exploring the api using api explorer page of your civi crm deployment

Okay, I see now. The explorer page is helpful. I can see that this is replacing documentation.
I think I will need to get access to our Civi CRM instance. I can't remember whether I was already given access. I can't find anything. Who should I contact, Rowena?

I might also need access to a web key so that I can at least reproduce the pipeline.

Reply:
Contact Rowena for access to our CiviCRM instance.
There is data hub user which Rowena has givne access to read the contacts data via api, and that is the only user whose `api_key`  we can use. Similarly, I have the site key as well and i can share both with you via slack